extensions/binding/org.eclipse.smarthome.binding.bluetooth.bluegiga/src/main/java/org/eclipse/smarthome/binding/bluetooth/bluegiga/internal/BlueGigaSerialHandler.java

```diff
@@ -0,0 +1,505 @@
+                removeTransactionListener(this);
+
+                return (T) response;
+            }
+
+            @Override
+            public boolean transactionEvent(BlueGigaResponse bleResponse) {
+                // Check if this response completes our transaction
+                if (bleCommand.hashCode() == bleResponse.hashCode()) {
+                    return false;
+                }
+
+                if (!expected.isInstance(bleResponse)) {
+                    // ignoring response if it was not requested
+                    logger.warn("Ignoring {} response which has not been requested.",
+                            bleResponse.getClass().getSimpleName());
+                    return false;
+                }
+
+                response = bleResponse;
+                complete = true;
+                synchronized (this) {
+                    notify();
+                }
+
+                return true;
+            }
+        }
+
+        Callable<T> worker = new TransactionWaiter();
+        return executor.submit(worker);
+    }
+
+    /**
+     * Sends a {@link BlueGigaCommand} request to the NCP and waits for the response. The response is correlated with
+     * the request and the returned {@link BlueGigaResponse} contains the request and response data.
+     *
+     * @param bleCommand {@link BlueGigaCommand}
+     * @return response {@link BlueGigaResponse}
+     */
+    public BlueGigaResponse sendTransaction(BlueGigaCommand bleCommand) {
+        checkIfAlive();
+        Future<BlueGigaResponse> futureResponse = sendBleRequestAsync(bleCommand, BlueGigaResponse.class);
+
+        try {
+            return futureResponse.get();
+        } catch (InterruptedException | ExecutionException e) {
+            throw new BlueGigaException("Error sending BLE transaction to listeners.", e);
+        }
+    }
+
+    /**
+     * Sends a {@link BlueGigaCommand} request to the NCP and waits for the response for specified period of time.
+     * The response is correlated with the request and the returned {@link BlueGigaResponse}
+     * contains the request and response data.
+     *
+     * @param bleCommand {@link BlueGigaCommand}
+     * @param timeout milliseconds to wait until {@link TimeoutException} is thrown
+     * @return response {@link BlueGigaResponse}
+     * @throws TimeoutException when specified timeout exceeds
+     */
+    public <T extends BlueGigaResponse> T sendTransaction(BlueGigaCommand bleCommand, Class<T> expected, long timeout)
+            throws TimeoutException {
+        Future<T> futureResponse = sendBleRequestAsync(bleCommand, expected);
+        try {
+            return futureResponse.get(timeout, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException | ExecutionException e) {
+            futureResponse.cancel(true);
+            throw new BlueGigaException("Error sending BLE transaction to listeners: ", e);
+        }
+    }
+
+    // TODO: Add a timeout mechanism
```

Comment:
Seeing the `TRANSACTION_TIMEOUT_PERIOD` parameter I guess this comment is now obsolete?